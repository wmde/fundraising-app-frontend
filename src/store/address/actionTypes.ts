export const validateAddress = 'validateAddress';
export const validateEmail = 'validateEmail';
export const setAddressField = 'setAddressField';
export const setAddressType = 'setAddressType';
export const setNewsletterOptIn = 'setNewsletterOptIn';
export const setReceiptOptOut = 'setReceiptOptOut';
