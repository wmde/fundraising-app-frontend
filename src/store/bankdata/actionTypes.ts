export const initializeBankData = 'initializeBankData';
export const markEmptyValuesAsInvalid = 'markEmptyFieldsAsInvalid';
export const markBankDataAsIncomplete = 'markBankDataAsIncomplete';
export const markBankDataAsInvalid = 'markBankDataAsInvalid';
export const setBankData = 'setBankData';
