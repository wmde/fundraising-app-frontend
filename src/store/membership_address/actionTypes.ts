export const validateAddress = 'validateAddress';
export const setAddressField = 'setAddressField';
export const setAddressType = 'setAddressType';
export const setEmail = 'setEmail';
export const setDate = 'setDate';
export const setNewsletterOptIn = 'setNewsletterOptIn';
export const setReceiptOptOut = 'setReceiptOptOut';
