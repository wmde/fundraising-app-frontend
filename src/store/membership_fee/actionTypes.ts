export const markEmptyFeeAsInvalid = 'markEmptyFeeAsInvalid';
export const markEmptyValuesAsInvalid = 'markEmptyValuesAsInvalid';
export const setFee = 'setFee';
export const setInterval = 'setInterval';
export const setType = 'setType';
