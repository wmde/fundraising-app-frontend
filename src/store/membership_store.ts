import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import createAddress from '@/store/membership_address';
import {
	NS_MEMBERSHIP_ADDRESS,
} from './namespaces';

Vue.use( Vuex );

export function createStore() {
	const storeBundle: StoreOptions<any> = {
		modules: {
			[ NS_MEMBERSHIP_ADDRESS ]: createAddress(),
		},
		strict: process.env.NODE_ENV !== 'production',
	};

	return new Vuex.Store<any>( storeBundle );
}
