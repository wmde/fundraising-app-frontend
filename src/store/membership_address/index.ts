import { Module } from 'vuex';
import { MembershipAddressState } from '@/view_models/Address';
import { Validity } from '@/view_models/Validity';
import { actions } from '@/store/membership_address/actions';
import { getters } from '@/store/membership_address/getters';
import { mutations } from '@/store/membership_address/mutations';
import { AddressTypeModel } from '@/view_models/AddressTypeModel';

export default function (): Module<MembershipAddressState, any> {
	const state: MembershipAddressState = {
		isValidating: false,
		addressType: AddressTypeModel.PERSON,
		receiptOptOut: false,
		values: {
			salutation: '',
			title: '',
			firstName: '',
			lastName: '',
			companyName: '',
			street: '',
			postcode: '',
			city: '',
			country: 'DE',
			email: '',
			date: '',
		},
		validity: {
			salutation: Validity.INCOMPLETE,
			title: Validity.VALID,
			firstName: Validity.INCOMPLETE,
			lastName: Validity.INCOMPLETE,
			companyName: Validity.INCOMPLETE,
			street: Validity.INCOMPLETE,
			postcode: Validity.INCOMPLETE,
			city: Validity.INCOMPLETE,
			country: Validity.VALID,
			email: Validity.INCOMPLETE,
			date: Validity.VALID,
			addressType: Validity.VALID,
		},
	};

	const namespaced = true;

	return {
		namespaced,
		state,
		getters,
		mutations,
		actions,
	};
}
