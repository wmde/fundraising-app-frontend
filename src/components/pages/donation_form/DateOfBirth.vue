<template>
    <fieldset>
		<label for="birthDate" class="subtitle has-margin-top-36">{{ $t( 'membership_birth_date_label' ) }}</label>
		<b-input type="text"
			id="birthDate"
			:placeholder="$t( 'membership_birth_date_placeholder' )"
			v-model="date"
			@blur="validateDate">
		</b-input>
		<span v-if="dateHasError" class="help is-danger">{{ $t( 'membership_birth_date_error' ) }}</span>
    </fieldset>
</template>

<script lang="ts">
import Vue from 'vue';
import { NS_MEMBERSHIP_ADDRESS } from '@/store/namespaces';
import { action } from '@/store/util';
import { setDate } from '@/store/membership_address/actionTypes';

export default Vue.extend( {
	name: 'DateOfBirth',
	data: function () {
		return {
			date: '',
			datePattern: new RegExp( [
				'^(?:(?:31(\\.)',
				'(?:0?[13578]|1[02]))\\1|(?:(?:29|30)(\\.)',
				'(?:0?[13-9]|1[0-2])\\2))(?:(?:1[6-9]|[2-9]\\d)?\\d{2})$|^(?:29(\\.)',
				'0?2\\3(?:(?:(?:1[6-9]|[2-9]\\d)?(?:0[48]|[2468][048]|[13579][26])|(?:(?:16|[2468][048]|[3579][26])00))))$',
				'|^(?:0?[1-9]|1\\d|2[0-8])(\\.)(?:(?:0?[1-9])|(?:1[0-2]))\\4(?:(?:1[6-9]|[2-9]\\d)?\\d{2})$',
			].join( '' ) ),
			dateHasError: false,
		};
	},
	methods: {
		setDate: function () {
			this.$store.dispatch( action( NS_MEMBERSHIP_ADDRESS, setDate ), this.$data.date );
		},
		validateDate: function () {
			this.dateHasError = !this.dateIsValid();
			if ( !this.dateHasError ) {
				this.setDate();
			}
		},
		dateIsValid: function () {
			return this.datePattern.test( this.$data.date );
		},
	},
} );
</script>
