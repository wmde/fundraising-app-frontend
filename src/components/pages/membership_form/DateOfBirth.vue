<template>
	<fieldset>
		<label for="birthDate" class="subtitle has-margin-top-36">{{ $t( 'membership_form_birth_date_label' ) }}</label>
		<b-field :type="{ 'is-danger': dateHasError }">
			<b-input type="text"
				id="birthDate"
				:placeholder="$t( 'membership_form_birth_date_placeholder' )"
				v-model="date"
				@blur="validateDate">
			</b-input>
		</b-field>
		<span v-if="dateHasError" class="help is-danger">{{ $t( 'membership_form_birth_date_error' ) }}</span>
	</fieldset>
</template>

<script lang="ts">
import Vue from 'vue';
import { NS_MEMBERSHIP_ADDRESS } from '@/store/namespaces';
import { action } from '@/store/util';
import { setDate } from '@/store/membership_address/actionTypes';

export default Vue.extend( {
	name: 'DateOfBirth',
	data: function () {
		return {
			date: '',
			datePattern: new RegExp( this.$props.validationPattern ),
			dateHasError: false,
		};
	},
	props: {
		validationPattern: String,
	},
	mounted() {
		this.$data.date = this.$store.state[ NS_MEMBERSHIP_ADDRESS ].values.date;
		if ( this.$data.date !== '' ) {
			this.validateDate();
		}
	},
	methods: {
		setDate: function () {
			this.$store.dispatch( action( NS_MEMBERSHIP_ADDRESS, setDate ), this.$data.date );
		},
		validateDate: function () {
			this.dateHasError = !this.dateIsValid();
			if ( !this.dateHasError ) {
				this.setDate();
			}
		},
		dateIsValid: function () {
			if ( this.$data.date === '' ) {
				return true;
			}
			return this.datePattern.test( this.$data.date );
		},
	},
} );
</script>
