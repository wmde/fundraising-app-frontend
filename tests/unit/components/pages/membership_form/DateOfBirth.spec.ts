import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Buefy from 'buefy';
import DateOfBirth from '@/components/pages/membership_form/DateOfBirth.vue';
import { createStore } from '@/store/membership_store';
import { NS_MEMBERSHIP_ADDRESS } from '@/store/namespaces';
import { action } from '@/store/util';
import { setDate } from '@/store/membership_address/actionTypes';

const localVue = createLocalVue();
localVue.use( Buefy );
localVue.use( Vuex );

describe( 'DateOfBirth.vue', () => {

	it( 'does not show an error on inital render', () => {
		const wrapper = mount( DateOfBirth, {
			localVue,
			store: createStore(),
			mocks: {
				$t: () => { },
			},
		} );
		expect( wrapper.vm.$data.dateHasError ).toBe( false );
	} );

	it( 'shows an error when the entered date is the wrong format', () => {
		const wrapper = mount( DateOfBirth, {
				localVue,
				store: createStore(),
				mocks: {
					$t: () => { },
				},
			} ),
			birthDate = wrapper.find( '#birthDate' );

		wrapper.setData( { date: '10-04-1975' } );
		birthDate.trigger( 'blur' );
		expect( wrapper.vm.$data.dateHasError ).toBe( true );

		wrapper.setData( { date: '10/04/1975' } );
		birthDate.trigger( 'blur' );
		expect( wrapper.vm.$data.dateHasError ).toBe( true );

		wrapper.setData( { date: '00.04.1975' } );
		birthDate.trigger( 'blur' );
		expect( wrapper.vm.$data.dateHasError ).toBe( true );

		wrapper.setData( { date: '10.13.1975' } );
		birthDate.trigger( 'blur' );
		expect( wrapper.vm.$data.dateHasError ).toBe( true );

		wrapper.setData( { date: '10.04.0042' } );
		birthDate.trigger( 'blur' );
		expect( wrapper.vm.$data.dateHasError ).toBe( true );
	} );

	it( 'sends the entered birth date to the store when it is correct', () => {
		const wrapper = mount( DateOfBirth, {
			localVue,
			store: createStore(),
			mocks: {
				$t: () => { },
			},
		} );
		const store = wrapper.vm.$store;
		store.dispatch = jest.fn().mockResolvedValue( null );
		const expectedAction = action( NS_MEMBERSHIP_ADDRESS, setDate );
		const birthDate = wrapper.find( '#birthDate' );

		wrapper.setData( { date: '09.11.1989' } );
		birthDate.trigger( 'blur' );
		let expectedPayload = '09.11.1989';
		expect( store.dispatch ).toBeCalledWith( expectedAction, expectedPayload );

		wrapper.setData( { date: '09/11/1989' } );
		birthDate.trigger( 'blur' );
		expectedPayload = '09/11/1989';
		expect( store.dispatch ).not.toBeCalledWith( expectedAction, expectedPayload );
	} );
} );
