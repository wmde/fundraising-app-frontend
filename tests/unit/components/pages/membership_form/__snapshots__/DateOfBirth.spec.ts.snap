// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateOfBirth.vue renders error message 1`] = `
<fieldset>
  <label
    class="subtitle has-margin-top-36"
    for="birthDate"
  >
    membership_form_birth_date_label
  </label>
  <div
    class="field"
  >
    <div
      class="control is-clearfix has-icons-right"
    >
      <input
        autocomplete="on"
        class="input is-danger"
        id="birthDate"
        name="birthDate"
        placeholder="membership_form_birth_date_placeholder"
        type="text"
      />
      <!--v-if-->
      <span
        class="icon is-right has-text-danger"
      >
        <i
          class="mdi mdi-alert-circle mdi-24px"
        />
      </span>
      <!--v-if-->
    </div>
  </div>
  <!--v-if-->
  <span
    class="help is-danger"
  >
    membership_form_birth_date_error
  </span>
</fieldset>
`;

exports[`DateOfBirth.vue renders field 1`] = `
<fieldset>
  <div
    class="field"
  >
    <div
      class="form-field form-field-text"
    >
      <label
        class="form-field-label"
        for="birthDate"
      >
        membership_form_birth_date_label
      </label>
      <div
        class="control text-form-input"
      >
        <input
          autocomplete="on"
          class="input"
          id="birthDate"
          name="birthDate"
          placeholder="membership_form_birth_date_placeholder"
          type="text"
        />
        <!--v-if-->
        <!--v-if-->
        <!--v-if-->
      </div>
      <!--v-if-->
      <span
        class="field-info-message"
      >
        
        
      </span>
    </div>
  </div>
  <span
    class="help has-text-dark-lighter"
  >
    membership_form_birth_date_help_text
  </span>
</fieldset>
`;
